Add `ImpalaHook`

blcksrx opened the pull request:
Description

Impala Hook

Use case / motivation

Related Issues
closes: #8916

uranusjr commented on setup.py:
```diff
@@ -293,6 +293,7 @@ def write_version(filename: str = os.path.join(*[my_dir, "airflow", "git_version
     'hmsclient>=0.1.0',
     'pyhive[hive]>=0.6.0',
 ]
+impala = ["impyla>=0.16.3"]
```
Any particular reasoning for this dependency range?

bdsoha replied:
@uranusjr I assume it has to do with peer-dependencies mentioned [on a previous PR](https://github.com/apache/airflow/pull/8918)

eladkal commented on airflow/providers/apache/impala/provider.yaml:
```diff
@@ -0,0 +1,35 @@
+    `Apache Impala <https://impala.apache.org/>`__
+
+versions:
+  - 1.0.0b2
```
Why beta?

eladkal commented on airflow/providers/apache/impala/README.md:
```diff
@@ -0,0 +1,59 @@
+ under the License.
+ -->
+
+# Package apache-airflow-backport-providers-apache-impala
```
This is not right. There is no backport provider.

I think this README file is redundant.

eladkal reviewed (changes requested):
We need a doc about how to setup connection. Check other peoviders for connection.rst examples

Commit:
impala hook

Update airflow/providers/apache/impala/hooks/impala.py

Co-authored-by: Dov Benyomin Sohacheski <b@kloud.email>

Update tests/providers/apache/impala/hooks/test_impala.py

Co-authored-by: Dov Benyomin Sohacheski <b@kloud.email>

Update tests/providers/apache/impala/hooks/test_impala.py

Co-authored-by: Dov Benyomin Sohacheski <b@kloud.email>

Update tests/providers/apache/impala/hooks/test_impala.py

Co-authored-by: Dov Benyomin Sohacheski <b@kloud.email>

impala connection docs and pytest

bdsoha commented on airflow/providers/apache/impala/hooks/impala.py:
```diff
@@ -0,0 +1,42 @@
+    conn_name_attr = 'impala_conn_id'
+    default_conn_name = 'impala_default'
+    conn_type = 'impala'
+    hook_name = "Impala"
```
You should replace all quotes with double-quotes in order to pass static checks.

```suggestion
class ImpalaHook(DbApiHook):
    """Interact with Apache Impala through impyla."""

    conn_name_attr = "impala_conn_id"
    default_conn_name = "impala_default"
    conn_type = "impala"
    hook_name = "Impala"
```

bdsoha commented on tests/providers/apache/impala/hooks/test_impala.py:
```diff
@@ -0,0 +1,110 @@
+        )
+
+        class UnitTestImapalHook(ImpalaHook):
+            conn_name_attr = 'impala_conn_id'
```
Replace quotes:

```suggestion
class TestImpalaHookConn(unittest.TestCase):
    def setUp(self):
        super().setUp()
        self.connection = Connection(
            login="login",
            password="password",
            host="host",
            port=21050,
            schema="test",
        )

        class UnitTestImapalHook(ImpalaHook):
            conn_name_attr = "impala_conn_id"
```

bdsoha commented on tests/providers/apache/impala/hooks/test_impala.py:
```diff
@@ -0,0 +1,110 @@
+        self.db_hook = UnitTestImpalaHook()
+
+    @patch('airflow.hooks.dbapi_hook.DbApiHook.insert_rows')
+    def test_insert_rows(self, mock_insert_rows):
```
Replace quotes:

```suggestion
        class UnitTestImpalaHook(ImpalaHook):
            conn_name_attr = "test_conn_id"

            def get_conn(self):
                return conn

        self.db_hook = UnitTestImpalaHook()

    @patch("airflow.hooks.dbapi_hook.DbApiHook.insert_rows")
    def test_insert_rows(self, mock_insert_rows):
```

bdsoha commented on tests/providers/apache/impala/hooks/test_impala.py:
```diff
@@ -0,0 +1,110 @@
+        result_sets = [('row1',), ('row2',)]
+        self.cur.description = [(column,)]
+        self.cur.fetchall.return_value = result_sets
+        df = self.db_hook.get_pandas_df(statement)
```
Replace quotes:

```suggestion
    def test_get_first_record(self):
        statement = "SQL"
        result_sets = [("row1",), ("row2",)]
        self.cur.fetchone.return_value = result_sets[0]

        self.assertEqual(result_sets[0], self.db_hook.get_first(statement))
        self.conn.close.assert_called_once_with()
        self.cur.close.assert_called_once_with()
        self.cur.execute.assert_called_once_with(statement)

    def test_get_records(self):
        statement = "SQL"
        result_sets = [("row1",), ("row2",)]
        self.cur.fetchall.return_value = result_sets

        self.assertEqual(result_sets, self.db_hook.get_records(statement))
        self.conn.close.assert_called_once_with()
        self.cur.close.assert_called_once_with()
        self.cur.execute.assert_called_once_with(statement)

    def test_get_pandas_df(self):
        statement = "SQL"
        column = "col"
        result_sets = [("row1",), ("row2",)]
        self.cur.description = [(column,)]
        self.cur.fetchall.return_value = result_sets
        df = self.db_hook.get_pandas_df(statement)
```

bdsoha reviewed (changes requested):
Since you opened this PR an additional check was added to *Static Checks* requiring the use of double-quotes.

eladkal commented on generated/provider_dependencies.json:
```diff
@@ -107,6 +107,12 @@
   },
+  "apache.impala": {
+    "deps": [
+      "impyla>=0.18.0,<1.0"
```
Why should we limit?
This invites questions when 1.0 will be released so if there is no good reason to limit in advanced we should relax the dependency.

blcksrx replied:
I believe it could broke some dependency as exactly what happened on old PR and we needed to wait for new version of `impyla`

eladkal commented on airflow/providers/apache/impala/provider.yaml:
```diff
@@ -0,0 +1,41 @@
+description: |
+    `Apache Impala <https://impala.apache.org/>`__
+versions:
+  - 2.4.1
```
```suggestion
  - 1.0.0
```

eladkal commented on airflow/providers/apache/impala/provider.yaml:
```diff
@@ -0,0 +1,41 @@
+  - 2.4.1
+
+dependencies:
+  - impyla>=0.18.0,<1.0
```
Should also have apache-airflow>=2.3.0

eladkal replied:
@potiuk should we consider to write a precommit that verify all providers have apache-airflow in provider yaml dependencies? I'm not sure because it seems like some providers don't have it 
https://github.com/apache/airflow/blob/8f5e100f30764e7b1818a336feaa8bb390cbb327/airflow/providers/http/provider.yaml#L38

eladkal replied:
@blcksrx this still need to be handled..
We need to add minimum Airflow version `apache-airflow>=2.3.0`

eladkal commented on tests/providers/apache/impala/hooks/test_impala.py:
```diff
@@ -0,0 +1,113 @@
+            schema="test",
+        )
+
+        class UnitTestImapalHook(ImpalaHook):
```
is it spelled right?

blcksrx replied:
Oh man no :(

eladkal commented on tests/providers/apache/impala/hooks/test_impala.py:
```diff
@@ -0,0 +1,113 @@
+        )
+
+
+class TestImpalaHook(unittest.TestCase):
```
Would be cool to write it as pytest
We have on going effort to do it so we can have less work in the future https://github.com/apache/airflow/pull/26831

blcksrx replied:
yeah sure

blcksrx replied:
done

Taragolis replied:
Class still based on `unittest.TestCase` and use their methods.
You need to:
1. Remove bases of `unittest.TestCase`
2. Rename `def setUp(self)` to `def setup_method`. And you don't need call `super()`
3. Instead of `self.assertEqual(a, b)` use just use `assert a == b`

eladkal reviewed (approved):
LGTM
we can merge when CI is green

bdsoha commented on INSTALL:
```diff
@@ -95,22 +95,15 @@ The list of available extras:
-# COMPILING FRONT-END ASSETS (in case you see "Please make sure to build the frontend in static/ directory and then restart the server")
-# Optional : Installing yarn - https://classic.yarnpkg.com/en/docs/install
-
-python setup.py compile_assets
```
@blcksrx It seems like these lines were accidentally removed.

blcksrx replied:
its not accidentally. last CI ran to a problem to fix it I needed to ran this command:
```shell
breeze setup regenerate-command-images
```
and it changed some files.
https://github.com/apache/airflow/actions/runs/3380995761/jobs/5614846732

uranusjr replied:
IIRC we merged `compile_assets` into startup so this is no longer needed? Don’t quite remember, I think @potiuk might?

potiuk replied:
Good eye @bdsoha - I think this was removed by accident early when generating extras (not really connected with regenerating images). Could you please restore it @blcksrx ?

@uranusjr: not really.

A bit of context.

The `compile_assets` is only automated when you start breeze, not in any other circumstances. The INSTALL here describes the process of installing airflow manualy from sources - this is the absolute requirement of the Apache Software Foundation, that "sufficient platform and tools" you should be able to take sources published in https://downloads.apache.org/airflow/2.4.3/ as *-sources.tar.gz and you should be able to get workinf software.

This is what INSTALL doc is for (it is also the first thing anyone who takes the .source.tgz file should look at). It is also described in this part of our installation documentation https://airflow.apache.org/docs/apache-airflow/stable/installation/index.html#using-released-sources. Those published sources should only contain source files that can and should be possible to manually edit to make any change to the software of ours. It should not contain any compiled sources 

This is described here: https://www.apache.org/legal/release-policy.html#source-packages

> Every ASF release MUST contain one or more source packages, which MUST be sufficient for a user to build and test the release provided they have access to the appropriate platform and tools. A source release SHOULD not contain compiled code.

This is not optional, it's a mandatory requirement of ASF release policy and we MUST follow it - otherwise PMC members of Apache Airlfow are not idemnified by the Apache Sofware Foundation in case of any legal battles - this is one of the reasons why we have release voting process and why PMC members only have binding votes.

In this context, running `python setup.py compile_assets` is absolutely necessary step in those instructions - we cannot (and we do not) have compiled assets released in the source.tgz. file and INSTALL document is about this file.

BTW. By answering this one I noticed that there are few things about installing providers and using constraints that are not needed here, actually, so I will shortly correct it.

blcksrx replied:
@potiuk done

eladkal commented on CONTRIBUTING.rst:
```diff
@@ -611,6 +611,7 @@ This is the full list of those extras:
 
   .. START EXTRAS HERE
 airbyte, alibaba, all, all_dbs, amazon, apache.atlas, apache.beam, apache.cassandra, apache.drill,
+<<<<<<< HEAD
```
You have a bad merge conflict

eladkal commented on INSTALL:
```diff
@@ -95,6 +95,7 @@ The list of available extras:
 
 # START EXTRAS HERE
 airbyte, alibaba, all, all_dbs, amazon, apache.atlas, apache.beam, apache.cassandra, apache.drill,
+<<<<<<< HEAD
```
Same

eladkal reviewed (changes requested):
The CI is stuck because of a bad merge

eladkal committed:
Merge branch 'main' into impala
